### PR TITLE
Feature: oft adapter step2 - bridge limits and fees in OFTAdapter

### PR DESCRIPTION
**Pull request by blueogin**
# Description

This PR inclues smart contracts which implementes bridge limits and fees in GoodDollarOFTAdapter.sol

About #7 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

**Review by sirpy — CHANGES_REQUESTED** [review]

**Review by sirpy — CHANGES_REQUESTED** [review]
There are no tests. If you would have added tests then you would see _enforcelimits isnt used

**Review by sirpy — CHANGES_REQUESTED** [review]

File: packages/bridge-contracts/contracts/oft/GoodDollarMinterBurner.sol [hunk 1]
```diff
@@ -66,14 +54,10 @@ contract GoodDollarMinterBurner is DAOUpgradeableContract {
 
 
     /**
-     * @dev Burn tokens from an address
+     * @dev Burn tokens from an address. Limits are not enforced on the sending side (burning); limits are enforced on the receiving side (minting).
```
Comment (sirpy) [inline packages/bridge-contracts/contracts/oft/GoodDollarMinterBurner.sol]: arent limits forced on both sides now?

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol [hunk 2]
```diff
@@ -68,6 +174,219 @@ contract GoodDollarOFTAdapter is OFTCoreUpgradeable {
...
+    /**
+     * @notice Function for pausing/unpausing the bridge
+     * @param _isPaused Whether to pause the bridge or not
+     */
+    function pauseBridge(bool _isPaused) external onlyOwner {
+        isClosed = _isPaused;
+        emit BridgePaused(_isPaused);
+    }
+
+    function approveFailedRequest(bytes32 _guid) external onlyOwner {
```
[the remark this answers is not in the document]
Comment (sirpy) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: owner should be able to approve bypassing optimistic window

**Review by sirpy — CHANGES_REQUESTED** [review]

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol [hunk 3]
```diff
@@ -68,6 +174,219 @@ contract GoodDollarOFTAdapter is OFTCoreUpgradeable {
...
+     * @param _amountLD The amount of tokens to credit in local decimals
+     * @return amountReceivedLD The amount of tokens actually received in local decimals
+     * @dev Fees are deducted on the destination chain
+     */
+    function _credit(
+        address _to,
+        uint256 _amountLD,
+        uint32 /* _srcEid */
+    ) internal virtual override returns (uint256 amountReceivedLD) {
+        if (_to == address(0x0)) _to = address(0xdead); // _mint(...) does not support address(0x0)
```
[the remark this answers is not in the document]
Comment (sirpy) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: where do you prevent address 0?

File: packages/bridge-contracts/contracts/oft/GoodDollarMinterBurner.sol (hunk 1 above)
Comment (blueogin) [reply packages/bridge-contracts/contracts/oft/GoodDollarMinterBurner.sol]: limits are forced on both sides now
it's just wrong comment
I will correct it

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (blueogin) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: Okay I will make it

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol (hunk 3 above)
[the remark this answers is not in the document]
Comment (blueogin) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: address 0 check is already available in GoodDollar(SuperGoodDollar)
you can check here:
https://github.com/GoodDollar/GoodProtocol/blob/b88e1566a1b6bb2c091e606a0608bc60ef3c443c/contracts/token/superfluid/SuperToken.sol#L276

[the remark this answers is not in the document]
Comment (sirpy) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: we need to prevent sending to address 0

[the remark this answers is not in the document]
Comment (blueogin) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: I will add an address-zero check in the _send function of GoodDollarOFTAdapter to prevent bridging to the zero address from the sender side.

**Comment by blueogin** [comment]
Hi @sirpy 
Could you please review?

**Comment by blueogin** [comment]
@sirpy 
Could you please review?

**Comment by blueogin** [comment]
I asked Cursor to audit smart contracts, and it identified missing zero-address checks and an incorrect transfer limit check
These issues have now been fixed.

File: packages/bridge-contracts/test/GoodDollarOFT.test.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+      await expect(minterBurner.connect(user).setOperator(adapter.address, true)).to.be.revertedWith(
+        "only avatar can call this method"
+      );
+
+      await expect(minterBurner.connect(avatar).setOperator(adapter.address, true))
+        .to.emit(minterBurner, "OperatorSet")
+        .withArgs(adapter.address, true);
+    });
+
+    it("operators can mint/burn; burn requires allowance", async () => {
```
Comment (sirpy) [inline packages/bridge-contracts/test/GoodDollarOFT.test.ts:92]: for permissions you ALWAYS need to test also the negative

**Comment by sirpy** [comment]
@copilot please review the code
1. add a github action that would run the unit tests on changes
2. are there any unit tests that you think are missing?

**Comment by sirpy** [comment]
@blueogin check the new unit tests by the AI, they are failing, what's harness it is using?

@copilot test should run on any pull request remove 
  branches:
      - feat/oft-adapter-step2
      - feat/oft-adapter-step1
      
  also modify the run command to run all test files in the test folder ie yarn hardhat run tests/

**Comment by blueogin** [comment]
@sirpy 
I reviewed changes by AI
the harness contract exists to make internal behavior testable - a helper contract used in unit tests to expose internals that are inaccessible from outside.
the failing tests are from very old existing tests - not from oft bridge

**Comment by blueogin** [comment]
@sirpy,
I’ve added a revert test case for the bridge limit on the sending side and updated the GitHub Actions workflow to run tests only for the OFT contracts.

Changes are ready for review.

**Comment by blueogin** [comment]
@sirpy, Could you please review

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol [hunk 5]
```diff
@@ -147,7 +147,9 @@ contract GoodDollarOFTAdapter is UUPSUpgradeable, OFTCoreUpgradeable {
         require(address(_minterBurner) != address(0), "minterBurner required");
         require(_owner != address(0), "owner required");
         require(_feeRecipient != address(0), "feeRecipient required");
-        require(IERC20Metadata(_token).decimals() == IERC20Metadata(_token).decimals(), 'token decimals mismatch');
+        uint8 sharedDecimals = OFTCoreUpgradeable.sharedDecimals();
+        uint8 tokenDecimals = IERC20Metadata(_token).decimals();
+        require(10 ** (tokenDecimals - sharedDecimals) == decimalConversionRate, 'token decimals mismatch');
```
Comment (sirpy) [inline packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: why is this required?
where is decimalConversionRate defined?

File: .github/workflows/bridge-contracts-tests.yml [hunk 6]
```diff
@@ -38,7 +38,7 @@ jobs:
         run: yarn build
         working-directory: packages/bridge-contracts
 
-      - name: Run all tests
```
Comment (sirpy) [inline .github/workflows/bridge-contracts-tests.yml]: all tests should run.
i've fixed the failing tests of other contracts

**Review by sirpy — CHANGES_REQUESTED** [review]

File: packages/bridge-contracts/test/GoodDollarOFT.test.ts (hunk 4 above)
Comment (blueogin) [reply packages/bridge-contracts/test/GoodDollarOFT.test.ts:92]: I will add negative permission checks

File: packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol (hunk 5 above)
Comment (blueogin) [reply packages/bridge-contracts/contracts/oft/GoodDollarOFTAdapter.sol]: In the constructor, the innerToken(param named as _token) address is used to calculate decimalConversionRate within OFTCoreUpgradeable based on that token’s decimals.
Later, the initialize method receives the innerToken address, so its decimals must match the token's decimals used during construction; otherwise, amount calculations in OFTCoreUpgradeable will break.

I made a small change to make it more direct:
Stored the constructor token’s decimals in innerTokenDecimals
Added a check in initialize to verify the provided token has matching decimals before proceeding

File: .github/workflows/bridge-contracts-tests.yml (hunk 6 above)
Comment (blueogin) [reply .github/workflows/bridge-contracts-tests.yml]: sure

**Comment by blueogin** [comment]
tests for other contracts are still failing @sirpy 
Is the fix on master branch?
